### PR TITLE
Migrate all core theming to vuex driven dynamic styles

### PR DESCRIPTION
**Pull request by rtibbles**
### Summary
#4524 did a lot of good things, but by separating each build into a separate, independent process, it broke the ability to do dynamic coreAPI injection at build time. Along with this, it also broke the ability to do custom styling of Kolibri at build time using a build time specified external plugin.

This PR rectifies some of this in a slightly different way by migrating all styling that hitherto has been done by variables specified in `core-theme.scss` to be driven by a core object in Vuex state.

As a bonus, it also adds a colour customization tool in dev mode to let people dynamically change styles.
…

### Reviewer guidance
Check that styling is still applied correctly in the many components that have been refactored - I still have some more work to do manually verifying all this also, although I was mostly checking as I went, and did a post-pass of high visibility areas.

Also read the additions to the docs and see if they make sense!

### References
No current issues this addresses, but will enable custom theming for accessibility and cosmetic purposes in future.

----

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Gherkin stories have been updated
- [ ] Unit tests have been updated

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.rst

### CODE REVIEW

File: docs/frontend_architecture/core.rst [hunk 1]
```diff
@@ -47,7 +47,39 @@ These modules would now be available for import anywhere with the following stat
...
+------------------
+
+In addition to some of the global SCSS variables, we also leverage Vuex state to drive overall theming of the application, in order to allow for more flexible theming (either for accessibility or cosmetic purposes). As such, much of our colour styles are defined in Javascript variables kept in Vuex state, which are then programmatically accessed using Vue.js style bindings in order to apply stylings inline to elements.
+
+This works for most use cases, however, inline styles cannot apply pseudo-selectors (e.g. ':hover', ':focus', '::before') - in order to accomplish this, we leverage the `Aphrodite<https://github.com/Khan/aphrodite>`__ library to generate computed classes that apply these stylings through a programmatically generated style sheet. This should be generally performant, but we are limiting their use for now to cases where we have pseudo selectors and styles that may conflict with pseudo-selector applied styles (to have clear selector precedence). In order to apply a style using Aphrodite, define a style object as a computed property, similarly to how you might for a Vue.js style binding. Pseudo-selectors can be encoded within this object thusly:
```
Comment (indirectlylit) [inline docs/frontend_architecture/core.rst]: Also for transitions.

One thing that's [been discussed lately](https://github.com/learningequality/kolibri/issues/4543) is distinguishing 'how-to' guides, references, and understanding-oriented discussions. If the goal here is mostly how-to, we could probably shorten it a bit.

For example:

> There are two cases where dynamic styles cannot be applied directly to DOM elements:
> 
> * styles using [pseudo-classes](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-classes)
> * styles applied during [Vue transitions](https://vuejs.org/v2/guide/transitions.html)
> 
> For these cases, it's necessary to define a "computed class" using the `$computedClass` function. This returns an auto-generated class name which can be used like a standard CSS class name. Under the hood, this uses [Aphrodite](https://github.com/Khan/aphrodite) to create unique classes for each set of inputs given, so be careful not to abuse this feature!

File: docs/frontend_architecture/core.rst [hunk 2]
```diff
@@ -47,7 +47,39 @@ These modules would now be available for import anywhere with the following stat
...
+
+.. code-block:: javascript
+
+  pseudoStyle() {
+    return {
+      ':hover': {
+        backgroundColor: white,
```
Comment (indirectlylit) [inline docs/frontend_architecture/core.rst]: let's actually reference a style here so it's clear what the point is

File: kolibri/core/assets/src/core-app/apiSpec.js [hunk 3]
```diff
@@ -181,14 +181,14 @@ export default {
...
   styles: {
-    theme,
     definitions,
     keenVars,
     buttonAndLinkStyles,
   },
   urls,
   utils: {
     contentNode,
+    colour,
```
Comment (indirectlylit) [inline kolibri/core/assets/src/core-app/apiSpec.js:191]: 🇬🇧

File: docs/frontend_architecture/core.rst (hunk 2 above)
Comment (rtibbles) [reply docs/frontend_architecture/core.rst]: Am I being very thick? Is `background-color` not a style?

Comment (indirectlylit) [reply docs/frontend_architecture/core.rst]: sorry I wasn't clear - `white` is not a vuex-driven style :)

Comment (rtibbles) [reply docs/frontend_architecture/core.rst]: Ah, gotcha. yes.

File: kolibri/core/assets/src/core-app/constructor.js [hunk 4]
```diff
@@ -64,6 +66,11 @@ export default class CoreApp {
       mediator.setReady();
     };
 
+    if (process.env.NODE !== 'production') {
+      const colourPicker = require('../utils/colourPicker').default;
+      colourPicker.start();
+    }
```
Comment (indirectlylit) [inline kolibri/core/assets/src/core-app/constructor.js:62]: This is fun, but I don't think it should be turned on by default in the devserver.

Also, `colourPicker.js` is unnecessary. You can just drop `ColourPicker.vue` in `CoreBase.vue`.

 Maybe add a prop on `CoreBase` like

```javascript
showColorPicker: {
  type: Boolean,
  default: false,
}
```

and then add to the template

```html
...

    <GlobalSnackbar />
    <ColourPicker v-if="showColorPicker" />
  </div>

</template>
```

File: kolibri/core/assets/src/views/CoreSnackbar/index.vue [hunk 5]
```diff
@@ -108,23 +108,12 @@
...
     margin: 16px;
-
-    /deep/ .ui-snackbar__action-button {
-      font-weight: bold;
-      color: $core-bg-light;
-    }
-
-    /deep/ .ui-button--type-secondary:hover:not(.is-disabled) {
-      background-color: rgba(255, 255, 255, 0.05);
-    }
```
Comment (indirectlylit) [inline kolibri/core/assets/src/views/CoreSnackbar/index.vue:127]: Looks like we lost some styles (regression of #2798):

image

**Review by indirectlylit — APPROVED** [review]
couple small comments. Overall I think it's best to get this in the mainline and open new issues as we find them

File: kolibri/core/assets/src/core-app/constructor.js (hunk 4 above)
Comment (rtibbles) [reply kolibri/core/assets/src/core-app/constructor.js:62]: Yeah, I did it like this so that it wouldn't even get bundled into the production build.

**Comment by indirectlylit** [comment]
getting some travis failures too

**Pull request merged**